### PR TITLE
Laboratory work #1, Amina Abdeyeva - 23FPL1

### PR DESCRIPTION
**Pull request by minabdeyeva**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -21,6 +21,25 @@ def tokenize(text: str) -> list[str] | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if text is None or type(text) is not str:
+        return None
+
+    tozenized_text = []
+    punctiation = ['!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '.', '-', '/', ':', ';', '<', '=', '>',
+                        '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~', ' ']
+    numbers = '0123456789'
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:30]: Can you store all "bad symbols" together? in one variable

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -21,6 +21,25 @@ def tokenize(text: str) -> list[str] | None:
...
+
+    for symb in text:
+        if symb in punctiation or symb in numbers:
+            continue
+        else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:35]: ``else` is unnecessary cause you use `continue`

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -21,6 +21,25 @@ def tokenize(text: str) -> list[str] | None:
...
+            low_symb = symb.lower()
+            tozenized_text.append(low_symb)
+
+    if tozenized_text == []:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:39]: bad condition. Use just `not tozenized_text`

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -35,6 +54,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if tokens is None or type(tokens) is not list:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:57]: `type(...) is not ...` should be replaced with `isinstance`

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -35,6 +54,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        return  None
+    frequencies = {}
+    for token in tokens:
+        if type(token) is not str:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:61]: The same - `isinstance`

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -35,6 +54,23 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+            return None
+        elif token in frequencies:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:63]: try to simplify - set 0 if `token not in frequencies` and do +1 always

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -49,7 +85,13 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
-
+    profile = {}
+    if type(language) is not str or language not in ['en', 'de'] or type(text) is not str:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:89]: THe same

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -49,7 +85,13 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+        return None
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:91]: ``else` is unnecessary cause you did `return`

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Comment by minabdeyeva** [comment]
done

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -21,6 +21,23 @@ def tokenize(text: str) -> list[str] | None:
...
+        return None
+
+    tozenized_text = []
+    bad_symbols = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '.', '-', '/', ':', ';', '<', '=', '>',
+                        '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~', ' ']
+
+    for symb in text:
+        if symb in bad_symbols:
+            continue
+        low_symb = symb.lower()
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:34]: let use `symb.lower()` only once for the source big string?

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -21,6 +21,23 @@ def tokenize(text: str) -> list[str] | None:
...
+        tozenized_text.append(low_symb)
+
+    if not tozenized_text:
+        return None
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:39]: `else` is unnecessary

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -49,7 +82,12 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
-
+    profile = {}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:85]: You can avoid creating the vatiable by using `return {'name': ..., ...}`

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -65,6 +103,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if not isinstance(predicted, list) or not isinstance(actual, list) or len(predicted) != len(actual):
+        return None
+    n = len(predicted)
+    mse = 0
+    for i in range(n):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:110]: use `enumerate`

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -65,6 +103,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     In case of corrupt input arguments, None is returned
     """
 
+    if not isinstance(predicted, list) or not isinstance(actual, list) or len(predicted) != len(actual):
+        return None
+    n = len(predicted)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:108]: I believe you don't need the variable

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -106,6 +174,27 @@ def detect_language(
     In case of corrupt input arguments, None is returned
     """
 
+    def sort_by_alpha(lang1, lang2):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:177]: No! Please, do not create functions inside functions!

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -106,6 +174,27 @@ def detect_language(
...
+        return None
+
+    score_1 = compare_profiles(unknown_profile, profile_1)
+    score_2 = compare_profiles(unknown_profile, profile_2)
+
+    if score_1 < score_2:
+        return profile_1['name']
+    elif score_1 == score_2:
+        return sort_by_alpha(profile_1['name'], profile_2['name'])
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:194]: `else` is unnecessary

File: seminars/practice_3_lists.py [hunk 16]
```diff
@@ -62,11 +62,19 @@ def count_evens(nums: list) -> int:
     """
     # student realization goes here
 
+    counter = 0
```
Comment (artyomtugaryov) [inline seminars/practice_3_lists.py:65]: revert the file.

File: lab_1_classify_by_unigrams/settings.json [hunk 17]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 8
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/settings.json:2]: 👎🏻

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
What about 10? I`m sure you can!

File: lab_1_classify_by_unigrams/settings.json (hunk 17 above)
Comment (minabdeyeva) [reply lab_1_classify_by_unigrams/settings.json:2]: done

**Review by artyomtugaryov — DISMISSED** [review]

**Pull request closed without merging**